### ISSUE TITLE
[JIT] Allows nested dictionaries to store Hyperparameters alongside models.

### ISSUE THREAD

**Issue by randolf-scholz** [pytorch/pytorch#71131]
### 🚀 The feature, motivation and pitch

Personally, I like to add hyperparameter dictionaries to my models, which helps to modularize them, like so:

```python
class ResNet(nn.ModuleList):
    DEFAULT_HP: Dict[str, Any] = {
        "__name__" : __qualname__,
        "__source__": __module__,
        "num_blocks" : 5,
        "block" : ResNetBlock.HP
    }
    """Default Hyperparameter choices.""""

    HP: dict[str, Any]
    """Actual Hyperparameter choices."""

    def __init__(self, input_size: int, **kwargs) - > None:
        super().__init__()
        self.HP = recursive_dict_update(deepcopy(self.DEFAULT_HP), **kwargs)
        self.HP["input_size"] = input_size
        self.HP["block"]["input_size"] = input_size
        
        blocks: list[nn.Module] = []

        for _ in range(self.HP["num_blocks"]):
             module: nn.Module = initialize_from_config(self.HP["block"])
             blocks.append(module)
       super().__init__(blocks)

    def forward(self, x: Tensor) -> Tensor:
        for block in self:
            x = x + block(x)
        return x
```

Here, `initialize_from_config` uses the fields `__name__` and `__source__` to automatically load and initialize the component-model. `ResNetBlock` itself might also have a nested config, e.g. which layers the block consists of. 

Now this all works fine and dandy until we try to `jit.script` such models. Ideally I would like to keep the selected Hyper-Parameter configuration as part of the model, so that they get serialized together when calling `jit.save` instead of having to json-dump it separately. 

The core issue is that the tracer does not play well with nested dictionary. With a bit of trial and error I found that it currently only works if the nested `dict`s always have consistent `value` type. (All the HP-dictionaries only use string keys and are guaranteed to be json-able).

So e.g. `HP: Dict[str, Any] = {"foo": None, "this" : {"is" : {"nested" : {"very" : "deep"}}}}` is fine, but `HP: Dict[str, Any] = {"nested" : {"foo" : "bar", "baz" : None}}` isn't

### Alternatives

One can obviously store Hyperparameters separately, but I think this approach is quite nice due to its modularity. Another advantage is that once we `jit.load` a model one could immediately reconstruct how it was initialized.
